### DISCUSSION TITLE
error handling & testing

### DISCUSSION THREAD

**Discussion by dtirer** [benbjohnson/wtf#11] [Q&A] [answered]
I was looking to see how you dealt with error handling, and more specifically, how you test that the correct errors are being returned.  I see that you check the error code, which could be common to a number of errors in a given function that you're testing, and then you check the end-user error message itself, which is specific to this particular error instance.  

```
// sqlite/dial_membership_test.go
t.Run("ErrDialRequired", func(t *testing.T) {
    db := MustOpenDB(t)
    defer MustCloseDB(t, db)
    s := sqlite.NewDialMembershipService(db)
    
    ctx := context.Background()
    _, ctx0 := MustCreateUser(t, ctx, db, &wtf.User{Name: "jane"})
    
    if err := s.CreateDialMembership(ctx0, &wtf.DialMembership{}); err == nil {
	    t.Fatal("expected error")
    } else if wtf.ErrorCode(err) != wtf.EINVALID || wtf.ErrorMessage(err) != `Dial required for membership.` {
	    t.Fatalf("unexpected error: %#v", err)
    }
})
```

This seems to make sense, however, I'm curious why you opted for this over creating sentinel errors in your domain package:

```
// dial_membership.go
var ErrDialRequired = errors.New("Dial required for membership")
```

And then your tests would check against that and wouldn't need to change as your end-user messages change. 

If the project were bigger, would you consider doing it differently or this still the approach you'd choose?

Thanks!

**Comment by benbjohnson** [comment, marked as answer]
Hi Daniel, great question. I used to use sentinel errors quite a bit but I've changed my approach to use code/message for a couple of reasons:

1. It works well with error wrapping—e.g. `fmt.Errorf("foo: %w, err)`. Ultimately I want to test what the end-user will see and not the wrapped error message (which I'll write to the log).

2. It works well over a network boundary. If you use parse an incoming error message from another server using `errors.New(s)` then that error won't `==` the sentinel error even though they have the same text.

3. Codes translate well to implementation-specific error codes (e.g. `http` or `grpc`).

**Reply by dtirer** [reply]
Okay thanks, that makes a lot of sense!

**Comment by marselester** [comment]
Hi Ben,

Thank you for sharing insights on how to structure Go projects! A few years ago I was researching this topic and I tried to incorporate your ideas as I understood them:

- "Failure is your Domain" https://github.com/marselester/ddd-err
- "Structuring Applications for Growth" https://marselester.com/how-to-structure-go-projects.html
